types/hooks.d.ts

```diff
@@ -0,0 +1,328 @@
+     */
+    functionAppDirectory: string;
+}
+
+/**
+ * Object passed to AppStartContext constructors.
+ * For testing purposes only
+ */
+export interface AppStartContextInit extends HookContextInit {
+    /**
+     * Defaults to 'unknown'
+     */
+    functionAppDirectory?: string;
+}
+
+/**
+ * Handler for app terminate hooks
+ */
+export type AppTerminateHandler = (context: AppTerminateContext) => void | Promise<void>;
+
+/**
+ * Context on a function app that is about to be terminated
+ * This object will be passed to all app terminate hooks
+ */
+export declare class AppTerminateContext extends HookContext {
+    /**
+     * For testing purposes only. This will always be constructed for you when run in the context of the Azure Functions runtime
+     */
+    constructor(init?: AppTerminateContextInit);
+}
+
+/**
+ * Object passed to AppTerminateContext constructors.
+ * For testing purposes only
+ */
+export interface AppTerminateContextInit extends HookContextInit {}
+
+/**
+ * Handler for pre-invocation hooks
+ */
+export type PreInvocationHandler = (context: PreInvocationContext) => void | Promise<void>;
+
+/**
+ * Base class for all invocation hook context objects (pre-invocation and post-invocation)
+ */
+export declare abstract class InvocationHookContext extends HookContext {
+    /**
+     * For testing purposes only. This will always be constructed for you when run in the context of the Azure Functions runtime
+     */
+    constructor(init?: InvocationHookContextInit);
+
+    /**
+     * The context object passed to the function
+     * This object is readonly. You may modify it, but attempting to overwrite it will throw an error
+     */
+    readonly invocationContext: InvocationContext;
+}
+
+/**
+ * Base interface passed to invocation hook context constructors.
+ * For testing purposes only
+ */
+interface InvocationHookContextInit extends HookContextInit {
+    /**
+     * Defaults to new InvocationContext with default values if not specified
+     */
+    invocationContext?: InvocationContext;
+}
+
+/**
+ * Context on a function that is about to be executed
+ * This object will be passed to all pre-invocation hooks
+ */
+export declare class PreInvocationContext extends InvocationHookContext {
+    /**
+     * For testing purposes only. This will always be constructed for you when run in the context of the Azure Functions runtime
+     */
+    constructor(init?: PreInvocationContextInit);
+
+    /**
+     * The arguments passed to this specific invocation.
+     * Changes to this array _will_ affect the inputs passed to your function
+     */
+    args: any[];
+
+    /**
+     * The function handler for this specific invocation. Changes to this value _will_ affect the function itself
+     */
+    functionHandler: FunctionHandler;
+}
+
+/**
+ * Object passed to PreInvocationContext constructors.
+ * For testing purposes only
+ */
+export interface PreInvocationContextInit extends InvocationHookContextInit {
```

Comment:
I feel like you told me this already, but I'm confused why we have both `PreInvocationContextInit` and `PreInvocationCoreContext`, with "core" nested into "init". Can we have just one object without duplication/nesting?